kaldb/src/main/java/com/slack/kaldb/clusterManager/CacheSlotAssignmentService.java

```diff
@@ -0,0 +1,253 @@
+    if (pendingTask == null || pendingTask.getDelay(TimeUnit.SECONDS) <= 0) {
+      pendingTask =
+          executorService.schedule(
+              this::assignCacheSlotsToReplicas,
+              managerConfig.getEventAggregationSecs(),
+              TimeUnit.SECONDS);
+    } else {
+      LOG.debug(
+          "Cache slot assignment already queued for execution, will run in {} ms",
+          pendingTask.getDelay(TimeUnit.MILLISECONDS));
+    }
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        managerConfig.getCacheSlotAssignmentServiceConfig().getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting cache slot assignment service");
+    cacheSlotMetadataStore.addListener(this::runOneIteration);
+    replicaMetadataStore.addListener(this::runOneIteration);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    executorService.shutdown();
+    LOG.info("Closed cache assignment service");
+  }
+
+  /**
+   * Assigns replicas to available cache slots, up to the configured replica lifespan min
+   * configuration. Replicas will be assigned with the most recently created first in descending
+   * order. In the event that more replicas than slots exist this ensures the most recent replicas
+   * are preferred. No preference is given to specific cache slots, which may result in over/under
+   * utilization of specific cache slots.
+   *
+   * <p>If this method fails to successfully assign all the replicas needing slot assignment, the
+   * following iteration of this method would attempt to re-assign these until there are no more
+   * available replicas to assign.
+   *
+   * @return The count of successfully assigned cache slots
+   */
+  protected int assignCacheSlotsToReplicas() {
+    Timer.Sample assignmentTimer = Timer.start(meterRegistry);
+
+    List<CacheSlotMetadata> availableCacheSlots =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            .filter(
+                cacheSlotMetadata ->
+                    cacheSlotMetadata.cacheSlotState.equals(
+                        Metadata.CacheSlotMetadata.CacheSlotState.FREE))
+            .collect(Collectors.toList());
+
+    List<String> assignedReplicaIds =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            .map(cacheSlotMetadata -> cacheSlotMetadata.replicaId)
+            .filter(replicaId -> !replicaId.isEmpty())
+            .collect(Collectors.toList());
+
+    List<String> replicaIdsToAssign =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            // only replicas created in the last X mins
+            .filter(
+                replicaMetadata ->
+                    replicaMetadata.createdTimeUtc
+                        > Instant.now()
```

Comment:
Moved timestamp into a variable.

I also reworked this so that we no longer have to use two separate filters. We now filter the list first with all of the criteria in one pass, so the following sort and mapping steps have a smaller set of items.